src/ILogger/ApplicationInsightsLoggerProvider.cs

```diff
@@ -0,0 +1,120 @@
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Text;
+    using Microsoft.ApplicationInsights;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Implementation;
+    using Microsoft.Extensions.Options;
+
+    /// <summary>
+    /// Application insights logger provider.
+    /// </summary>
+    /// <seealso cref="Microsoft.Extensions.Logging.ILoggerProvider" />
+    /// <seealso cref="Microsoft.Extensions.Logging.ISupportExternalScope" />
+    [ProviderAlias("ApplicationInsights")]
+    public class ApplicationInsightsLoggerProvider : ILoggerProvider, ISupportExternalScope
+    {
+        /// <summary>
+        /// The application insights logger options.
+        /// </summary>
+        private readonly ApplicationInsightsLoggerOptions applicationInsightsLoggerOptions;
+
+        /// <summary>
+        /// The telemetry client to be used to log messages to Application Insights.
+        /// </summary>
+        private readonly TelemetryClient telemetryClient;
+
+        /// <summary>
+        /// The collection of application insights loggers stored against categoryName.
+        /// categoryName -> <see cref="ApplicationInsightsLogger"/> map.
+        /// </summary>
+        private readonly ConcurrentDictionary<string, ApplicationInsightsLogger> applicationInsightsLoggers
+            = new ConcurrentDictionary<string, ApplicationInsightsLogger>();
+
+        /// <summary>
+        /// The external scope provider to allow setting scope data in messages.
+        /// </summary>
+        private IExternalScopeProvider externalScopeProvider;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApplicationInsightsLoggerProvider"/> class.
+        /// </summary>
+        /// <param name="telemetryConfiguration">The telemetry configuration.</param>
+        /// <param name="applicationInsightsLoggerOptions">The application insights logger options.</param>
+        /// <exception cref="System.ArgumentNullException">
+        /// telemetryConfiguration
+        /// or
+        /// loggingFilter
+        /// or
+        /// applicationInsightsLoggerOptions.
+        /// </exception>
+        public ApplicationInsightsLoggerProvider(
+            TelemetryConfiguration telemetryConfiguration,
+            IOptions<ApplicationInsightsLoggerOptions> applicationInsightsLoggerOptions)
+        {
+            if (telemetryConfiguration == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryConfiguration));
+            }
+
+            this.applicationInsightsLoggerOptions = applicationInsightsLoggerOptions?.Value ?? throw new ArgumentNullException(nameof(applicationInsightsLoggerOptions));
+
+            this.telemetryClient = new TelemetryClient(telemetryConfiguration);
```

Comment:
Ah I see. What happens if there are 2 of these in the system. Is there still a single collector instance shipping logs to an http backend?

Reply:
No in that case 2 telemetry clients will individually ship logs to AI. The equation however is not always this simple and there is a whole world of TelemetrySinks which can make everything sync or Async ordered etc. (I have recently started exploring the depths of AI SDK so my knowledge is still limited).

So in short if you are using everything default, TelemetryClients will be shipping their individual entries to AI opening 2 connections. This is effectively no different than 2 micro-services using the same AI key running on same machine.

Reply:
So there will be 2 background threads in ASP.NET Core applications that use this logger and enable the AI site extension right?

Reply:
Yes, 2 telemetry clients exist and these 2 telemetry clients are used in different scopes. One if used in the scope of startup.cs while the other one is used for the rest of the pipelines (requests etc.). These can emit the logs to (potentially) AI instances. This is however not a feature, but a side effect.